src/patterns/security-code/index.md.njk

```diff
@@ -0,0 +1,111 @@
+---
+title: Security code
+description: Identifying users when they sign in
+section: Patterns
+theme: Ask users for…
+aliases: 2FA
+backlog_issue_id: 25
+layout: layout-pane.njk
+---
+
+{% from "_example.njk" import example %}
+
+Use this pattern to verify users via a security code sent to their phone.
+
+{{ example({group: "patterns", item: "security-code", example: "default", html: true, nunjucks: true, open: false}) }}
+
+## When to use this pattern
+
+Use this pattern if your users need to sign in to your service.
+
+Requiring users to enter a security code sent to their mobile phone provides a second layer of security over a [password](/patterns/passwords/).
+
+You can require a security code on every sign in or only once per device, depending upon the risk of your service.
+
+## How it works
+
+You should first ask users for their [password](/patterns/passwords/). Once they have successfully entered this, send a text message to their mobile phone with a 5 digit code in this format:
```

Comment:
We probably need a bit more evidence for 5 digits. The default (Google, Apple, authenticator apps) is 6 digits. However I reckon that fewer digits means fewer mistakes, and in other contexts 4 digits are used, for example the default PIN length for a bank card.

When we were researching the sign up journey for Notify we had a participant with dyslexia who commented that 5 digits would be the most they thought they could transcribe correctly first time.

When @joelanman was working on GovWifi he found that avoiding consecutive repeated digits (e.g. 11223) helped people with dyslexia.

Anything which reduces the possible number of codes is theoretically making the code easier to guess. But I’d be interested to work out what the smallest number that practically prevents guessing is, when used in combination with other measures (e.g. throttling, account locking).

The aim should be for good enough security but great usability.

Reply:
@quis good to know, thanks! We based our decision to use a 5 digit code purely on the basis of copying Notify... 😄  The guidance _could_ be vague on the exact number of digits to use , but I think it’s more helpful to give a definite number if we can.

Reply:
I would say:
- start with 4 digits (like the PIN number on a bank card)
- only use more if you don’t have other technical measures to protect against automated attacks

I’ve used a couple of commercial services recently (AirBNB, Uber Eats) which have used 4-digit codes.

A friend with dyscalculia commented recently that they find the 6 digits on the Google Authenticator app really difficult.

I’m going to see if I can get someone from NCSC to weigh in on the technical side of this.

Reply:
Guidance on the ‘other measures’ that we should link to: https://www.ncsc.gov.uk/collection/passwords/updating-your-approach#PasswordGuidance:UpdatingYourApproach-Usethrottlingoraccountlockout

Reply:
My concern about 4 is that it might make interoperability with authenticator apps more difficult. Most authenticator apps are configured by reading a QR code that encodes a `otpauth://` URL and the [`otpauth://` URL format](https://github.com/google/google-authenticator/wiki/Key-Uri-Format) only allows for 6 or 8 digits to be specified (and some authenticator apps ignore that and assume it’s always 6).

Using 4 digits for codes sent over SMS but 6 digits for codes from an authenticator app is a possibility, of course.

Reply:
I checked the analytics data for Notify, which uses a 5 digit code with no spaces. Users miskey the code about 1 in every 50 attempts. I expect some users would make up a higher proportion of those attempts than others.

Reply:
Great discussion. I'd be up for reducing it to 4 digits, with some guidance around other security measures.

Reply:
@quis I've reduced the recommended length to 4 digits in https://github.com/alphagov/govuk-design-system/pull/1265/commits/edf257413fad4aa0fe1aab387067350ed3746a28, and added a summary of the technical recommendations from NCSC (with a link). Let me know what you think.

Reply:
That commit looks good 👍🏻